common/src/main/java/com/discordsrv/common/feature/bansync/BanSyncModule.java

```diff
@@ -206,13 +203,29 @@ public void onAccountLinked(AccountLinkedEvent event) {
         }
     }
 
-    private CompletableFuture<Guild.@Nullable Ban> getBan(Guild guild, long userId) {
-        return guild.retrieveBan(UserSnowflake.fromId(userId)).submit().exceptionally(t -> {
-            if (t instanceof ErrorResponseException && ((ErrorResponseException) t).getErrorResponse() == ErrorResponse.UNKNOWN_BAN) {
-                return null;
-            }
-            throw (RuntimeException) t;
-        });
+    private CompletableFuture<@Nullable Punishment> getBanOrBanRoled(Guild guild, long userId, BanSyncConfig config) {
+        UserSnowflake snowflake = UserSnowflake.fromId(userId);
+        return guild.retrieveBan(snowflake)
+                .submit()
+                .handle((ban, t) -> {
+                    if (t != null) {
+                        if (t instanceof ErrorResponseException && ((ErrorResponseException) t).getErrorResponse() == ErrorResponse.UNKNOWN_BAN) {
+                            // Not banned, but they might still have some of the banned roles
+                            return guild.retrieveMember(snowflake)
+                                    .submit()
+                                    .handle((member, throwable) -> {
+                                        if (throwable instanceof ErrorResponseException && ((ErrorResponseException) throwable).getErrorResponse() == ErrorResponse.UNKNOWN_MEMBER) return null;
+
+                                        if (member.getRoles().stream().anyMatch(role -> config.minecraftToDiscord.role.roleId == role.getIdLong())) {
+                                            return new Punishment(null, null, null);
+                                        }
+
+                                        return null;
+                                    });
+                        }
+                        throw (RuntimeException) t;
+                    } else return CompletableFuture.completedFuture(this.punishment(ban));
```

Comment:
```suggestion
                    } else {
                        return CompletableFuture.completedFuture(this.punishment(ban));
                    }
```

Alternatively some of the extra indentation could be eliminated by moving this to the top and not using else